### PR TITLE
Jk/subtotal

### PR DESCRIPTION
**Pull request by izulin**
### Context

14 new functions: VAR.P, VAR.S, VARA, VARPA, STDEV.P, STDEV.S, STDEVA, STDEVPA, SUBTOTAL, PRODUCT, VARP, VAR, STDEVP, STDEV.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature or improvement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Additional language file or change to the existing one (translations)

### Related issue(s):
1.
2.
3.

### Checklist:

- [x] My code follows the code style of this project,
- [x] My change requires a change to the documentation,
- [x] I described the modification in the CHANGELOG.md file.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -10,6 +10,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Added 9 text functions EXACT, LOWER, UPPER, MID, T, SUBSTITUTE, REPLACE, UNICODE, UNICHAR. (#159)
 - Added 5 datetime functions: INTERVAL, NETWORKDAYS, NETWORKDAYS.INTL, WORKDAY, WORKDAY.INTL. (#153)
 - Added 3 information functions HLOOKUP, ROW, COLUMN. (#158)
+- Added 8 statistical functions VAR.P, VAR.S, VARA, VARPA, STDEV.P, STDEV.S, STDEVA, STDEVPA.
+- Added 1 mathemtical function SUBTOTAL.
```
Comment (voodoo11) [inline CHANGELOG.md:14]:
```suggestion
- Added 1 mathematical function SUBTOTAL.
```

**Review by voodoo11 — APPROVED** [review]

File: test/interpreter/function-varp.spec.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+  it('should calculate variance (population)', () => {
+    const engine = HyperFormula.buildFromArray([
+      ['=VAR.P(2, 3)'],
+      ['=VAR.P(B2:D2, E2, F2)', 2, 3, 'foo', 'bar', 4],
+    ])
+    expect(engine.getCellValue(adr('A1'))).toEqual(0.25)
+    expect(engine.getCellValue(adr('A2'))).toBeCloseTo(2/3, 6)
+  })
+})
```
Comment (wojciechczerniak) [inline test/interpreter/function-varp.spec.ts:22]: Some tests missing:

- If only one number is provided, returns 0.
- Logical values, and text representations of numbers that you type directly into the list of arguments are counted.
- If an argument is an array or reference, only numbers in that array or reference are counted. Empty cells, logical values, text, or error values in the array or reference are ignored.
- Arguments that are error values or text that cannot be translated into numbers cause errors.
- If you want to include logical values and text representations of numbers in a reference as part of the calculation, use the VARPA function.

Source: 
https://support.microsoft.com/en-us/office/var-p-function-73d1285c-108c-4843-ba5d-a51f90656f3a
https://docs.oasis-open.org/office/OpenDocument/v1.3/csprd02/part4-formula/OpenDocument-v1.3-csprd02-part4-formula.html#__RefHeading__1018790_715980110

File: test/interpreter/function-vara.spec.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+  it('should calculate variance (sample)', () => {
+    const engine = HyperFormula.buildFromArray([
+      ['=VARA(2, 3)'],
+      ['=VARA(B2:D2, E2, F2)', 2, 3, 'foo', 'bar', 4],
+    ])
+    expect(engine.getCellValue(adr('A1'))).toEqual(0.5)
+    expect(engine.getCellValue(adr('A2'))).toEqual(3.2)
+  })
+})
```
Comment (wojciechczerniak) [inline test/interpreter/function-vara.spec.ts:24]: Missing tests:

- Arguments can be the following: numbers; names, arrays, or references that contain numbers; text representations of numbers; or logical values, such as TRUE and FALSE, in a reference.
- Logical values and text representations of numbers that you type directly into the list of arguments are counted.
- Arguments that contain TRUE evaluate as 1; arguments that contain text or FALSE evaluate as 0 (zero).
- If an argument is an array or reference, only values in that array or reference are used. Empty cells and text values in the array or reference are ignored.
- Arguments that are error values or text that cannot be translated into numbers cause errors.
- If you do not want to include logical values and text representations of numbers in a reference as part of the calculation, use the VAR function.


Source:
https://docs.oasis-open.org/office/OpenDocument/v1.3/csprd02/part4-formula/OpenDocument-v1.3-csprd02-part4-formula.html#__RefHeading__1018788_715980110
https://support.microsoft.com/en-us/office/vara-function-3de77469-fa3a-47b4-85fd-81758a1e1d07

File: test/interpreter/function-stdevs.spec.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+  it('should calculate standard deviation (sample)', () => {
+    const engine = HyperFormula.buildFromArray([
+      ['=STDEV.S(2, 3)'],
+      ['=STDEV.S(B2:D2, E2, F2)', 2, 3, 'foo', 'bar', 4],
+    ])
+    expect(engine.getCellValue(adr('A1'))).toBeCloseTo(0.707106781186548, 6)
+    expect(engine.getCellValue(adr('A2'))).toEqual(1)
+  })
+})
```
Comment (wojciechczerniak) [inline test/interpreter/function-stdevs.spec.ts:24]: Missing tests:

- Arguments can either be numbers or names, arrays, or references that contain numbers.
- Logical values and text representations of numbers that you type directly into the list of arguments are counted.
- If an argument is an array or reference, only numbers in that array or reference are counted. Empty cells, logical values, text, or error values in the array or reference are ignored.
- Arguments that are error values or text that cannot be translated into numbers cause errors.
- If you want to include logical values and text representations of numbers in a reference as part of the calculation, use the STDEVA function.

https://support.microsoft.com/en-us/office/stdev-s-function-7d69cf97-0c1f-4acf-be27-f3e83904cc23

File: test/interpreter/function-stdevpa.spec.ts [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+  it('should calculate standard deviation (population)', () => {
+    const engine = HyperFormula.buildFromArray([
+      ['=STDEVPA(2, 3)'],
+      ['=STDEVPA(B2:D2, E2, F2)', 2, 3, 'foo', 'bar', 4],
+    ])
+    expect(engine.getCellValue(adr('A1'))).toEqual(0.5)
+    expect(engine.getCellValue(adr('A2'))).toBeCloseTo(1.6, 6)
+  })
+})
```
Comment (wojciechczerniak) [inline test/interpreter/function-stdevpa.spec.ts:22]: Please add missing tests and make it obvious that those arguments are supported or not:

- Arguments can be the following: numbers; names, arrays, or references that contain numbers; text representations of numbers; or logical values, such as TRUE and FALSE, in a reference.
- Text representations of numbers that you type directly into the list of arguments are counted.
- Arguments that contain TRUE evaluate as 1; arguments that contain text or FALSE evaluate as 0 (zero).
- If an argument is an array or reference, only values in that array or reference are used. Empty cells and text values in the array or reference are ignored.
- Arguments that are error values or text that cannot be translated into numbers cause errors.
- If you do not want to include logical values and text representations of numbers in a reference as part of the calculation, use the STDEVP function.

We should mark the requirements and the differences between functions clearly 
Source: https://support.microsoft.com/en-us/office/stdevpa-function-5578d4d6-455a-4308-9991-d405afe2c28c

> Unlike the STDEV function, includes values of type Text and Logical. Text values are treated as number 0. Logical TRUE is treated as 1, and FALSE is treated as 0. Empty cells are not included.
>
> In the sequence, only Numbers and Logical types are considered; cells with Text are converted to 0; other types are ignored. If Logical types are a distinct type, they are still included, with TRUE considered 1 and FALSE considered 0. Any Sample may be of type ReferenceList.
>
> The handling of string constants as parameters is implementation-defined. Either, string constants are converted to numbers, if possible and otherwise, they are treated as zero, or string constants are always treated as zero.

https://docs.oasis-open.org/office/OpenDocument/v1.3/csprd02/part4-formula/OpenDocument-v1.3-csprd02-part4-formula.html#__RefHeading__1018772_715980110

File: test/interpreter/function-stdevp.spec.ts [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+  it('should calculate standard deviation (population)', () => {
+    const engine = HyperFormula.buildFromArray([
+      ['=STDEV.P(2, 3)'],
+      ['=STDEV.P(B2:D2, E2, F2)', 2, 3, 'foo', 'bar', 4],
+    ])
+    expect(engine.getCellValue(adr('A1'))).toEqual(0.5)
+    expect(engine.getCellValue(adr('A2'))).toBeCloseTo(0.816496580927726, 6)
+  })
+})
```
Comment (wojciechczerniak) [inline test/interpreter/function-stdevp.spec.ts:22]: Tests are missing as in other cases but here I have to ask which version of STDEV.P we implement?

ODFF has STDEVP only, but Excel has both STDEVP and STDEV.P. One is improved, other for backward compatibility. How does that relate to the ODFF one? ODFF has the old one or already the improved version?

File: test/interpreter/function-stdeva.spec.ts [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+  it('should calculate standard deviation (sample)', () => {
+    const engine = HyperFormula.buildFromArray([
+      ['=STDEVA(2, 3)'],
+      ['=STDEVA(B2:D2, E2, F2)', 2, 3, 'foo', 'bar', 4],
+    ])
+    expect(engine.getCellValue(adr('A1'))).toBeCloseTo(0.707106781186548, 6)
+    expect(engine.getCellValue(adr('A2'))).toBeCloseTo(1.78885438199983, 6)
+  })
+})
```
Comment (wojciechczerniak) [inline test/interpreter/function-stdeva.spec.ts:24]: https://docs.oasis-open.org/office/OpenDocument/v1.3/csprd02/part4-formula/OpenDocument-v1.3-csprd02-part4-formula.html#__RefHeading__1018768_715980110

> **Unlike the STDEV function includes values of type Text and Logical. Text values are treated as number 0. Logical TRUE is treated as 1, and FALSE is treated as 0. Empty cells are not included.**
>
> The handling of string constants as parameters is implementation-defined. Either, string constants are converted to numbers, if possible and otherwise, they are treated as 0, or string constants are always treated as 0.

🙈  Which implementation did we choose? How does it relate to Excel requirements? 


- Arguments can be the following: numbers; names, arrays, or references that contain numbers; text representations of numbers; or logical values, such as TRUE and FALSE, in a reference.
- Arguments that contain TRUE evaluate as 1; arguments that contain text or FALSE evaluate as 0 (zero).
- If an argument is an array or reference, only values in that array or reference are used. Empty cells and text values in the array or reference are ignored.
- Arguments that are error values or text that cannot be translated into numbers cause errors.
- If you do not want to include logical values and text representations of numbers in a reference as part of the calculation, use the STDEV function.

Source: https://docs.oasis-open.org/office/OpenDocument/v1.3/csprd02/part4-formula/OpenDocument-v1.3-csprd02-part4-formula.html#__RefHeading__1018768_715980110

BTW We should add more tests for the requirements above

**Review by wojciechczerniak — CHANGES_REQUESTED** [review]
We should add to https://github.com/handsontable/hyperformula/blob/master/docs/guide/known-limitations.md that SUBTOTAL is not supported fully. There is no difference between `1` and `101`. We don't have support for row/column visibility meta data. @izulin @voodoo11 is this even possible? A callback in options maybe?

```ts
hfInstance = new Hyperformula({
  isHidden: function(rowIndex) {
    return handsontableInstance.isHidden(rowIndex);
  }
});
```

File: src/i18n/languages/csCZ.ts [hunk 8]
```diff
@@ -153,14 +153,15 @@ const dictionary: RawTranslationPackage = {
     OR: 'NEBO',
     PI: 'PI',
     PMT: 'PLATBA',
-    SUBSTITUTE: 'DOSADIT',
+    PRODUCT: 'SOUČIN',
```
[the remark this answers is not in the document]
Comment (izulin) [reply src/i18n/languages/csCZ.ts:156]: Right, thanks for catching this!

File: test/interpreter/function-stdeva.spec.ts (hunk 7 above)
Comment (izulin) [reply test/interpreter/function-stdeva.spec.ts:24]: > Which implementation did we choose? How does it relate to Excel requirements?

I don't really understand -- we implemented STDEVA function. The logic behind -A variants are nothing new, we already handled it with AVERAGEA, MAXA, MINA, SUMA functions.

File: test/interpreter/function-stdevp.spec.ts (hunk 6 above)
Comment (izulin) [reply test/interpreter/function-stdevp.spec.ts:22]: The STDEV.P is just a newer version of STDEVP and STDEV functions of excel. There is no difference between them.

**Comment by izulin** [comment]
> We should add to https://github.com/handsontable/hyperformula/blob/master/docs/guide/known-limitations.md that SUBTOTAL is not supported fully. There is no difference between `1` and `101`. We don't have support for row/column visibility meta data. @izulin @voodoo11 is this even possible? A callback in options maybe?
> 
> ```ts
> hfInstance = new Hyperformula({
>   isHidden: function(rowIndex) {
>     return handsontableInstance.isHidden(rowIndex);
>   }
> });
> ```

That would be super messy with the logic behind triggering recalculations, caching etc. (if done correctly, would probably result in performance hit).

File: test/interpreter/function-stdeva.spec.ts (hunk 7 above)
Comment (wojciechczerniak) [reply test/interpreter/function-stdeva.spec.ts:24]: ODFF description suggests that STDEV can have different behavior depending on the implementation: "The handling of string constants as parameters is implementation-defined. Either, string constants are converted to numbers, if possible and otherwise, they are treated as 0, or string constants are always treated as 0."

**Comment by wojciechczerniak** [comment]
> That would be super messy with the logic behind triggering recalculations, caching etc. (if done correctly, would probably result in performance hit).

So we have two known issues with SUBTOTAL, it can't ignore other SUBTOTALs and there is no difference between options `1` and `101`.

File: test/interpreter/function-stdeva.spec.ts (hunk 7 above)
Comment (izulin) [reply test/interpreter/function-stdeva.spec.ts:24]: I see. Then for consistency I would keep it the same as in AVERAGEA. Where should it be documented?

File: test/interpreter/function-stdevp.spec.ts (hunk 6 above)
Comment (wojciechczerniak) [reply test/interpreter/function-stdevp.spec.ts:22]: Cool, then nothing to do. Pitty we don't have function name aliases. We could map STDEVP to STDEV.P and be compatible with ODFF and Excel

**Comment by izulin** [comment]
> > That would be super messy with the logic behind triggering recalculations, caching etc. (if done correctly, would probably result in performance hit).
> 
> So we have two known issues with SUBTOTAL, it can't ignore other SUBTOTALs and there is no difference between options `1` and `101`.

I mentioned one of those in known issues, and other is documented in a testfile.

File: test/interpreter/function-product.spec.ts [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+describe('Function PRODUCT', () => {
+  it('should take at least one argument', () => {
+    const engine = HyperFormula.buildFromArray([
+      ['=PRODUCT()']
+    ])
+
+    expect(engine.getCellValue(adr('A1'))).toEqual(detailedError(ErrorType.NA, ErrorMessage.WrongArgNumber))
+  })
+
+  it('should calculate product', () => {
```
Comment (bardek8) [inline test/interpreter/function-product.spec.ts:14]: probably another relevant test would be with `PRODUCT(5,"foo")`

File: test/interpreter/function-stdeva.spec.ts [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+    expect(engine.getCellValue(adr('A2'))).toEqual(detailedError(ErrorType.DIV_BY_ZERO))
+  })
+
+  it('should calculate standard deviation (sample)', () => {
+    const engine = HyperFormula.buildFromArray([
+      ['=STDEVA(2, 3)'],
+      ['=STDEVA(B2:I2, 2, 3, 4, TRUE(), FALSE(), "1",)', 2, 3, 4, true, false, 'a', '\'1', null],
+    ])
+    expect(engine.getCellValue(adr('A1'))).toBeCloseTo(0.707106781186548, 6)
+    expect(engine.getCellValue(adr('A2'))).toBeCloseTo(1.50639661750509, 6)
```
Comment (bardek8) [inline test/interpreter/function-stdeva.spec.ts:22]: Please mark inconsistency with GS.

**Review by bardek8 — CHANGES_REQUESTED** [review]
As for the tests, you mostly covered what wojciechczerniak asked about, but I have few tiny issues:
- could you add tests for variance which accepts 1 argument? Just to make sure, that this corner-case is handled correctly.
- as for the coercions and white characters (thanks for the big test!), it turns out that all the functions that take into account the size of sample/population, they behave differently in XL and GS, once treating FALSE() as an element and once not. Can we have one big test with all coercions that behave similarly on XL and GS, and one that highlights the difference? Currently we stick to the behavior of XL, which is fine, but I believe that a test highlighting the inconsistency would be very useful.

File: CHANGELOG.md [hunk 11]
```diff
@@ -10,6 +10,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Added 9 text functions EXACT, LOWER, UPPER, MID, T, SUBSTITUTE, REPLACE, UNICODE, UNICHAR. (#159)
 - Added 5 datetime functions: INTERVAL, NETWORKDAYS, NETWORKDAYS.INTL, WORKDAY, WORKDAY.INTL. (#153)
 - Added 3 information functions HLOOKUP, ROW, COLUMN. (#158)
+- Added 8 statistical functions VAR.P, VAR.S, VARA, VARPA, STDEV.P, STDEV.S, STDEVA, STDEVPA.
+- Added 2 mathematical function SUBTOTAL, PRODUCT.
```
Comment (izulin) [inline CHANGELOG.md:14]:
```suggestion
- Added 2 mathematical functions SUBTOTAL, PRODUCT.
```

**Review by bardek8 — APPROVED** [review]
Thanks for the fixes!
LGTM, but let's wait for a green light from @wojciechczerniak

File: test/interpreter/function-subtotal.spec.ts [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+      ['=SUBTOTAL(12345, A2:A4, A5)'],
+      [3],
+      [5],
+      [2],
+      [4]
+    ])
+
+    expect(engine.getCellValue(adr('A1'))).toEqual(detailedError(ErrorType.VALUE, ErrorMessage.BadMode))
+  })
+})
```
[the remark this answers is not in the document]
Comment (wojciechczerniak) [reply test/interpreter/function-subtotal.spec.ts:174]: Known limitation.

File: test/interpreter/function-vars.spec.ts [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+  it('should calculate variance (sample)', () => {
+    const engine = HyperFormula.buildFromArray([
+      ['=VAR.S(2, 3)'],
+      ['=VAR.S(2, 3, 4, TRUE(), FALSE(), "1",)'],
+      ['=VAR.S(B3:I3)', 2, 3, 4, true, false, 'a', '\'1', null],
+    ])
+    expect(engine.getCellValue(adr('A1'))).toEqual(0.5)
+    expect(engine.getCellValue(adr('A2'))).toBeCloseTo(2.28571428571429, 6) //inconsistency with product #1
+    expect(engine.getCellValue(adr('A3'))).toEqual(1)
+  })
```
Comment (wojciechczerniak) [inline test/interpreter/function-vars.spec.ts:25]: Please add clear descriptions to the conditions we're checking

```
it('should ignore empty cells, logical values, text and error values if argument is a reference', () => {
      ['=VAR.S(B3:I3)', 2, 3, 4, true, false, 'a', '\'1', null, '#DIV/0'],
```

```
it('should count logical values and text representation of numbers that are arguments', () => {
      ['=VAR.S(2, 3, 4, TRUE(), FALSE(), "1",)'],
```

File: test/interpreter/function-varpa.spec.ts [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+
+    expect(engine.getCellValue(adr('A1'))).toEqualError(detailedError(ErrorType.NA, ErrorMessage.WrongArgNumber))
+  })
+
+  it('should calculate variance (population)', () => {
+    const engine = HyperFormula.buildFromArray([
+      ['=VARPA(2, 3)'],
+      ['=VARPA(2, 3, 4, TRUE(), FALSE(), "1",)'],
+      ['=VARPA(B3:I3)', 2, 3, 4, true, false, 'a', '\'1', null],
+      ['=VARPA(1)'],
```
Comment (wojciechczerniak) [inline test/interpreter/function-varpa.spec.ts:19]: Please split into clear descriptions that we're testing. From the previous comment it can be split into:

```
it('should ignore empty values or text that cannot be coerced to number in references', () => {
it('should evaluate TRUE as 1 and FALSE as 0 (zero)', () => {
```

Any others?

File: test/interpreter/function-varp.spec.ts [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+      ['=VAR.P()'],
+    ])
+
+    expect(engine.getCellValue(adr('A1'))).toEqualError(detailedError(ErrorType.NA, ErrorMessage.WrongArgNumber))
+  })
+
+  it('should calculate variance (population)', () => {
+    const engine = HyperFormula.buildFromArray([
+      ['=VAR.P(2, 3)'],
+      ['=VAR.P(2, 3, 4, TRUE(), FALSE(), "1",)'],
```
Comment (wojciechczerniak) [inline test/interpreter/function-varp.spec.ts:17]: `['=VAR.P(1)']` does not say anything about the condition. `0` is expected when `1` is an argument? 
```
it('should return zero if only one number is provided', () => {
it('should ignore empty cells, logical values, text or error values if they're a part of reference', () => {

```

File: test/interpreter/function-vara.spec.ts [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+    expect(engine.getCellValue(adr('A1'))).toEqualError(detailedError(ErrorType.NA, ErrorMessage.WrongArgNumber))
+    expect(engine.getCellValue(adr('A2'))).toEqualError(detailedError(ErrorType.DIV_BY_ZERO))
+  })
+
+  it('should calculate variance (sample)', () => {
+    const engine = HyperFormula.buildFromArray([
+      ['=VARA(2, 3)'],
+      ['=VARA(2, 3, 4, TRUE(), FALSE(), "1",)'],
+      ['=VARA(B3:I3)', 2, 3, 4, true, false, 'a', '\'1', null],
+    ])
```
Comment (wojciechczerniak) [inline test/interpreter/function-vara.spec.ts:21]:
```
it('should evaluate TRUE as 1, FALSE as 0', () => {
it('should ignore empty cells, text if they're a part of reference, logical values are counted', () => {
```
Maybe something else we can describe better?

**Review by wojciechczerniak — CHANGES_REQUESTED** [review]
I've added few examples but all tests should be reviewed if they can be improved.

File: test/interpreter/function-varp.spec.ts (hunk 15 above)
Comment (izulin) [reply test/interpreter/function-varp.spec.ts:17]: Zero from 1 arg is not any special case, its just part of how variance works. The only valid description is something like 'accepts one argument'.

File: test/interpreter/function-vars.spec.ts (hunk 13 above)
Comment (izulin) [reply test/interpreter/function-vars.spec.ts:25]: Errors are not ignored.

Comment (wojciechczerniak) [reply test/interpreter/function-vars.spec.ts]: Inconsistency with the docs then? Not again...

File: docs/guide/known-limitations.md [hunk 17]
```diff
@@ -34,4 +34,6 @@ you can't compare the arguments in a formula like this:
   * Asynchronous functions
   * Structured references ("Tables")
   * Currency data type
-  * Relative named expressions
\ No newline at end of file
+  * Relative named expressions
+ * Functions cannot use UI metadata (e.g. hidden rows for SUBTOTAL).
+ * SUBTOTAL function does not ignore nested subtotals.
```
Comment (wojciechczerniak) [inline docs/guide/known-limitations.md:39]:
```suggestion
  * Relative named expressions
  * Functions cannot use UI metadata (e.g. hidden rows for SUBTOTAL).
  * SUBTOTAL function does not ignore nested subtotals.
```

File: test/interpreter/function-varpa.spec.ts [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+  })
+
+  it('should coerce explicit argument to numbers', () => {
+    const engine = HyperFormula.buildFromArray([
+      ['=VARPA(2, 3, 4, TRUE(), FALSE(), "1",)'],
+    ])
+    expect(engine.getCellValue(adr('A1'))).toBeCloseTo(1.95918367346939, 6)
+  })
+
+  it('should ignore non-numeric values in ranges', () => {
```
Comment (wojciechczerniak) [inline test/interpreter/function-varpa.spec.ts:30]: VAR.P has the same description and the same values but different result. I think that the description is not precise enough. One includes logical values the other dosn't.

From "If you do not want to include logical values and text representations of numbers in a reference as part of the calculation, use the VARP function." I would say that here w should mention that this INCLUDES logical values and text representation

```suggestion
  it('should evaluate TRUE to 1, FALSE to 0 and coerce text representation of numbers', () => {
```

File: test/interpreter/function-varp.spec.ts [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+  })
+
+  it('should coerce explicit argument to numbers', () => {
+    const engine = HyperFormula.buildFromArray([
+      ['=VAR.P(2, 3, 4, TRUE(), FALSE(), "1",)'],
+    ])
+    expect(engine.getCellValue(adr('A1'))).toBeCloseTo(1.95918367346939, 6) //inconsistency with product #1
+  })
+
+  it('should ignore non-numeric values in ranges', () => {
```
Comment (wojciechczerniak) [inline test/interpreter/function-varp.spec.ts:30]: VAR.P has the same description and the same values but different result. I think that the description is not precise enough. One includes logical values the other dosn't.

After "If you want to include logical values and text representations of numbers in a reference as part of the calculation, use the VARPA function." I would go with something like:

```suggestion
  it('should ignore non-numeric values in ranges, including ignoring logical values and text representation of numbers', () => {
```

File: test/interpreter/function-vara.spec.ts [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+  })
+
+  it('should coerce explicit argument to numbers', () => {
+    const engine = HyperFormula.buildFromArray([
+      ['=VARA(2, 3, 4, TRUE(), FALSE(), "1",)'],
+    ])
+    expect(engine.getCellValue(adr('A1'))).toBeCloseTo(2.28571428571429)
+  })
+
+  it('should ignore non-numeric values in ranges', () => {
```
Comment (wojciechczerniak) [inline test/interpreter/function-vara.spec.ts:30]: Same situtation with VARA and VARS as I've described for VARPS and VARPA.

```suggestion
  it('should evaluate TRUE to 1, FALSE to 0 and coerce text representation of numbers', () => {
```

File: test/interpreter/function-vars.spec.ts [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+
+
+  it('should coerce explicit argument to numbers', () => {
+    const engine = HyperFormula.buildFromArray([
+      ['=VAR.S(2, 3, 4, TRUE(), FALSE(), "1",)'],
+    ])
+    expect(engine.getCellValue(adr('A1'))).toBeCloseTo(2.28571428571429, 6) //inconsistency with product #1
+  })
+
+  it('should ignore non-numeric values in ranges', () => {
```
Comment (wojciechczerniak) [inline test/interpreter/function-vars.spec.ts:31]: Same situtation with VARA and VARS as I've described for VARPS and VARPA.

```suggestion
  it('should ignore non-numeric values in ranges, including ignoring logical values and text representation of numbers', () => {
```

**Review by wojciechczerniak — CHANGES_REQUESTED** [review]
Awesome! Few descriptions can be more specific though. They say the same but results are different. I've added specifics to two of them but could you apply those to all pairs?
STDEVA vs STDEVS
STDEVPA vs STDEVP
VARA vs VARS
VARPA vs VARP

File: test/interpreter/function-varp.spec.ts (hunk 15 above)
Comment (wojciechczerniak) [reply test/interpreter/function-varp.spec.ts]: 👍

File: src/interpreter/plugin/NumericAggregationPlugin.ts [hunk 22]
```diff
@@ -91,6 +108,36 @@ export class NumericAggregationPlugin extends FunctionPlugin {
...
     },
+    'PRODUCT': {
+      method: 'product',
+    },
+    'VAR.S': {
+      method: 'vars',
+    },
+    'VAR.P': {
+      method: 'varp',
+    },
```
Comment (wojciechczerniak) [inline src/interpreter/plugin/NumericAggregationPlugin.ts:119]: Would be a valid solution to make VARP and VARS aliases here that are linking to the same method but under different name?

```suggestion
    'VAR.S': {
      method: 'vars',
    },
    'VAR.P': {
      method: 'varp',
    },
    'VARS': {
      method: 'vars',
    },
    'VARP': {
      method: 'varp',
    },
```

BTW. Are we missing arguments definitions or they are redundant?

File: test/interpreter/function-stdeva.spec.ts (hunk 7 above)
Comment (wojciechczerniak) [reply test/interpreter/function-stdeva.spec.ts:43]: I would create a test that is our behaviour documentation with a clear description and an inconsistency comment

File: src/interpreter/plugin/NumericAggregationPlugin.ts (hunk 22 above)
Comment (izulin) [reply src/interpreter/plugin/NumericAggregationPlugin.ts:119]: those functions are not using the mechanism that requires args definitions

File: test/interpreter/function-vars.spec.ts (hunk 13 above)
Comment (izulin) [reply test/interpreter/function-vars.spec.ts]: Where is it in the docs?

File: src/interpreter/plugin/NumericAggregationPlugin.ts (hunk 22 above)
Comment (izulin) [reply src/interpreter/plugin/NumericAggregationPlugin.ts:119]: There is no such method as 'VARS' -- VAR should be an alias of VAR.S

File: test/interpreter/function-vars.spec.ts (hunk 13 above)
Comment (wojciechczerniak) [reply test/interpreter/function-vars.spec.ts]: "Empty cells, logical values, text, or **error values** in the array or reference are ignored." Source: https://support.microsoft.com/en-us/office/var-s-function-913633de-136b-449d-813e-65a00b2b990b

Comment (izulin) [reply test/interpreter/function-vars.spec.ts]: Well, the implementation clearly propagates errors 🥇

File: src/interpreter/plugin/NumericAggregationPlugin.ts (hunk 22 above)
Comment (wojciechczerniak) [reply src/interpreter/plugin/NumericAggregationPlugin.ts:119]: Ok, we have a problem on our list
https://github.com/handsontable/formula-parser/blob/d056ea2351153b77831b9a92842ef198b3f517f0/src/supported-formulas.js#L371-L376

File: test/interpreter/function-vars.spec.ts (hunk 13 above)
Comment (wojciechczerniak) [reply test/interpreter/function-vars.spec.ts]: Ok, nothing to add here then 🤷 Thanks for checking

**Review by wojciechczerniak — APPROVED** [review]
Great! 🥇
